File: app/src/main/java/me/fru1t/worddropper/widget/colortheme/ColoredLinearLayout.java
package me.fru1t.worddropper.widget.colortheme;

import android.content.Context;
import android.support.annotation.Nullable;
import android.util.AttributeSet;
import android.widget.LinearLayout;

import me.fru1t.worddropper.R;
import me.fru1t.worddropper.WordDropperApplication;
import me.fru1t.worddropper.settings.colortheme.ColorThemeViewProxy;
import me.fru1t.worddropper.settings.colortheme.ColorThemeXml;

/**
 * An automatically colored LinearLayout
 */
public class ColoredLinearLayout extends LinearLayout {
    private final WordDropperApplication app;
    private final ColorThemeViewProxy proxy;

    public ColoredLinearLayout(Context context) {
        this(context, null, 0);
    }

    public ColoredLinearLayout(Context context, @Nullable AttributeSet attrs) {
        this(context, attrs, 0);
    }

    public ColoredLinearLayout(Context context, @Nullable AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        app = (WordDropperApplication) context.getApplicationContext();
        proxy = new ColorThemeViewProxy(this, attrs, R.styleable.ColoredLinearLayout,
                new ColorThemeViewProxy.AttributeMap(
                        R.styleable.ColoredLinearLayout_backgroundColorTheme,
                        ColorThemeXml.BACKGROUND,
                        this::setBackgroundColor));
    }

    @Override
    protected void onDetachedFromWindow() {
        super.onDetachedFromWindow();
        app.removeColorThemeEventHandler(proxy);
    }

    @Override
    protected void onAttachedToWindow() {
        super.onAttachedToWindow();
        app.addColorThemeEventHandler(proxy);
    }
}

File: app/src/main/java/me/fru1t/worddropper/ui/widget/ColoredLinearLayout.kt
package me.fru1t.worddropper.ui.widget

import android.content.Context
import android.util.AttributeSet
import android.widget.LinearLayout
import me.fru1t.android.slik.Slik
import me.fru1t.android.slik.annotations.Inject
import me.fru1t.worddropper.R
import me.fru1t.worddropper.WordDropperApplication
import me.fru1t.worddropper.settings.colortheme.AttributeMap
import me.fru1t.worddropper.settings.colortheme.ColorThemeViewProxy
import me.fru1t.worddropper.settings.colortheme.ColorThemeViewProxyFactory
import me.fru1t.worddropper.settings.colortheme.ColorThemeXml

/** An automatically colored LinearLayout  */
open class ColoredLinearLayout @JvmOverloads constructor(
        context: Context, attrs: AttributeSet? = null, defStyleAttr: Int = 0)
    : LinearLayout(context, attrs, defStyleAttr) {
    private @Inject lateinit var factory: ColorThemeViewProxyFactory
    private val proxy: ColorThemeViewProxy

    init {
        Slik.get(WordDropperApplication::class).inject(this, ColoredLinearLayout::class)
        proxy = factory.create(
                this,
                attrs,
                R.styleable.ColoredLinearLayout,
                AttributeMap(
                        R.styleable.ColoredLinearLayout_backgroundColorTheme,
                        ColorThemeXml.BACKGROUND,
                        { setBackgroundColor(it) }))
    }
}
